language-plutus-core/src/Language/PlutusCore/Normalize/Internal.hs

```diff
@@ -157,30 +159,18 @@ are added to environments and normalization instantiates all variables presented
 -- | Normalize a 'Type' in the 'NormalizeTypeM' monad.
 normalizeTypeM
     :: (HasUnique (tyname ann) TypeUnique, MonadQuote m)
-    => Type tyname ann -> NormalizeTypeT m tyname ann (Normalized (Type tyname ann))
-normalizeTypeM (TyForall ann name kind body) =
-    TyForall ann name kind <<$>> normalizeTypeM body
-normalizeTypeM (TyIFix ann pat arg)          =
-    TyIFix ann <<$>> normalizeTypeM pat <<*>> normalizeTypeM arg
-normalizeTypeM (TyFun ann dom cod)           =
-    TyFun ann <<$>> normalizeTypeM dom <<*>> normalizeTypeM cod
-normalizeTypeM (TyLam ann name kind body)    =
-    TyLam ann name kind <<$>> normalizeTypeM body
-normalizeTypeM (TyApp ann fun arg)           = do
-    vFun <- normalizeTypeM fun
-    vArg <- normalizeTypeM arg
-    case unNormalized vFun of
-        TyLam _ nArg _ body -> do
-            countTypeNormalizationStep
-            substNormalizeTypeM vArg nArg body
-        _                   -> pure $ TyApp ann <$> vFun <*> vArg
-normalizeTypeM var@(TyVar _ name)            = do
-    mayTy <- lookupTyNameM name
-    case mayTy of
-        Nothing -> pure $ Normalized var
-        Just ty -> liftDupable ty
-normalizeTypeM builtin@TyBuiltin{}           =
-    pure $ Normalized builtin
+    => Type tyname ann
+    -> NormalizeTypeT m tyname ann (Normalized (Type tyname ann))
+normalizeTypeM t = Normalized <$> transformMOf typeSubtypes norm t where
+    norm (TyApp _ (TyLam _ argName _ body) arg) = do
+        countTypeNormalizationStep
+        unNormalized <$> substNormalizeTypeM (Normalized arg) argName body
+    norm var@(TyVar _ name) = do
+        mayTy <- lookupTyNameM name
+        case mayTy of
+            Nothing -> pure var
+            Just ty -> unNormalized <$> liftDupable ty
+    norm x = pure x
```

Comment:
I very much prefer the previous version as it is quite more readable to me and deals with the `Normalized` wrapper much more explicitly. E.g. you introduced `Normalized arg`, but in order to understand whether it's safe to use the constructor I need to understand the exact operational behavior of `transformMOf`. I think the previous version was much clearer and safer due to having fewer `Normalized` and `unNormalized` calls.

Reply:
Can we just pattern match on `TyApp` and `TyVar` here keeping all the interesting cases explicit and handle the rest using the new generic machinery?

Reply:
So the problem again is that the traversals are type-preserving - they don't work with `Type -> NormalizedType`, only `Type -> Type`. They let us look at our `Type`s as containers of `Type`s, but not of other things.

So I had to change the main function here to be `Type -> Type`, which is what you don't like.